sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/BlobAsyncClient.java

```diff
@@ -420,51 +417,66 @@ as we can guarantee we only need at most two buffers for any call to write (two
-            AtomicLong totalProgress = new AtomicLong(0);
-            Lock progressLock = new ReentrantLock();
+        final ParallelTransferOptions finalParallelTransferOptions = parallelTransferOptions == null
+            ? new ParallelTransferOptions()
+            : parallelTransferOptions;
 
+        try {
             return Mono.using(() -> uploadFileResourceSupplier(filePath),
                 channel -> {
-                    final SortedMap<Long, String> blockIds = new TreeMap<>();
-                    return Flux.fromIterable(sliceFile(filePath, finalParallelTransferOptions.getBlockSize()))
-                        .doOnNext(chunk -> blockIds.put(chunk.getOffset(), getBlockID()))
-                        .flatMap(chunk -> {
-                            String blockId = blockIds.get(chunk.getOffset());
-
-                            Flux<ByteBuffer> progressData = ProgressReporter.addParallelProgressReporting(
-                                FluxUtil.readFile(channel, chunk.getOffset(), chunk.getCount()),
-                                progressReceiver, progressLock, totalProgress);
-
-                            return getBlockBlobAsyncClient()
-                                .stageBlockWithResponse(blockId, progressData, chunk.getCount(), null);
-                        })
-                        .then(Mono.defer(() -> getBlockBlobAsyncClient().commitBlockListWithResponse(
-                            new ArrayList<>(blockIds.values()), headers, metadata, tier, accessConditions)))
-                        .then()
-                        .doOnTerminate(() -> {
-                            try {
-                                channel.close();
-                            } catch (IOException e) {
-                                throw logger.logExceptionAsError(new UncheckedIOException(e));
-                            }
-                        });
+                    try {
+                        BlockBlobAsyncClient blockBlobAsyncClient = getBlockBlobAsyncClient();
+                        long fileSize = channel.size();
+
+                        // If the file is larger than 256MB chunk it and stage it as blocks.
+                        if (fileSize > 256 * Constants.MB) {
+                            return uploadBlocks(fileSize, finalParallelTransferOptions, headers, metadata, tier,
+                                accessConditions, channel, blockBlobAsyncClient);
+                        } else {
+                            // Otherwise we know it can be sent in a single request reducing network overhead.
+                            return blockBlobAsyncClient.uploadWithResponse(FluxUtil.readFile(channel), fileSize,
+                                headers, metadata, tier, accessConditions)
+                                .then();
+                        }
+                    } catch (IOException ex) {
+                        return Mono.error(new UncheckedIOException(ex));
```

Comment:
Why is the `IOException` wrapped with `UncheckedIOException`? Can this just be `Mono.error(ex)`?

Reply:
Changed